Mathlib/FieldTheory/JacobsonNoether.lean

```diff
@@ -0,0 +1,215 @@
+      ⟨expChar_pow_pos D p m, by rw [hm2 (p ^ m + 1) (Nat.le_add_right _ _)]; rfl⟩⟩
+    classical
+    refine ⟨Nat.find h_exist, ⟨(Nat.find_spec h_exist).1, ?_, (Nat.find_spec h_exist).2⟩⟩
+    set t := (Nat.find h_exist - 1 : ℕ) with ht
+    by_cases h_pos : 0 < t
+    · convert (ne_eq _ _) ▸ not_and.mp (Nat.find_min h_exist (m := t) (by omega)) h_pos
+      omega
+    · suffices h_find: Nat.find h_exist = 1 by
+        rwa [h_find, pow_one]
+      rw [not_lt, Nat.le_zero, ht, Nat.sub_eq_zero_iff_le] at h_pos
+      linarith [(Nat.find_spec h_exist).1]
+  -- We define `c` to be the value that we proved above to be non-zero.
+  set c := ((δ a) ^ n) b with hc_def
+  letI : Invertible c := ⟨c⁻¹, inv_mul_cancel₀ hb.1, mul_inv_cancel₀ hb.1⟩
+  -- We prove that `c` commutes with `a`.
+  have hc : a * c = c * a := by
+    apply eq_of_sub_eq_zero
+    rw [← mulLeft_apply (R := k), ← mulRight_apply (R := k),
+      ← δ_def, δ_iterate_succ a b n, hb.2]
+  -- We now make some computation to obtain the final equation.
+  set d := c⁻¹ * a * ((δ a) ^ (n - 1)) b with hd_def
+  have hc': c⁻¹ * a = a * c⁻¹ := by
+    apply_fun (c⁻¹ * · * c⁻¹) at hc
+    rw [mul_assoc, mul_assoc, mul_inv_cancel₀ hb.1, mul_one, ← mul_assoc,
+      inv_mul_cancel₀ hb.1, one_mul] at hc
+    exact hc
+  have c_eq : a * ((δ a) ^ (n - 1)) b - ((δ a) ^ (n - 1)) b * a = c := by
+    rw [hc_def, ← Nat.sub_add_cancel hn, ← δ_iterate_succ, δ_def]; rfl
+  have eq1 : c⁻¹ * a * ((δ a)^ (n - 1)) b - c⁻¹ * ((δ a) ^ (n - 1)) b * a = 1 := by
+    simp_rw [mul_assoc, (mul_sub_left_distrib c⁻¹ _ _).symm, c_eq, inv_mul_cancel_of_invertible]
+  -- We show that `a` commutes with `d`.
+  have deq : a * d - d * a = a := by
+    nth_rw 3 [← mul_one a]
+    rw [hd_def, ← eq1, mul_sub, mul_assoc _ _ a, sub_right_inj, hc',
+      ← mul_assoc, ← mul_assoc, ← mul_assoc]
+  -- This then derives a contradiction.
+  apply_fun (a⁻¹ * · ) at deq
+  rw [mul_sub, ← mul_assoc, inv_mul_cancel₀ ha₀, one_mul, ← mul_assoc, sub_eq_iff_eq_add] at deq
+  obtain ⟨r, hr⟩ := exists_pow_mem_center_of_inseparable p d insep
+  apply_fun (· ^ (p ^ r)) at deq
+  rw [add_pow_expChar_pow_of_commute p r (Commute.one_left _) , one_pow,
+    ← DivisionSemiring.conj_pow ha₀, ← hr.comm, mul_assoc, inv_mul_cancel₀ ha₀, mul_one,
+    self_eq_add_left] at deq
+  exact one_ne_zero deq
+
+open Subring Algebra in
+/-- Jacobson-Noether theorem: For a non-commutative algebraic
+  division algebra `D` (with base ring being a field `L`), if the center of
+  `D` over `L` is `L`, then there exist an element `x` of `D \ L`
+  that is separable over `L`. -/
+theorem exists_separable_mem_of_not_central' {L D : Type*} [Field L] [DivisionRing D]
+    [Algebra L D] [Algebra.IsAlgebraic L D]
+    (hcenter : Subalgebra.center L D = ⊥) (hneq : (⊥ : Subalgebra L D) ≠ ⊤) :
+    ∃ x : D, x ∉ (⊥ : Subalgebra L D) ∧ IsSeparable L x := by
+  have ntrivial : center D ≠ ⊤ :=
+    congr(Subalgebra.toSubring $hcenter).trans_ne (Subalgebra.toSubring_injective.ne hneq)
```

Comment:
Can you explain (to me) this syntax with `congr`, no space before the `(` and the `$`?

Reply:
This part was done by @FR-vdash-bot as I mentioned in commit [bb0c028](https://github.com/leanprover-community/mathlib4/pull/16525/commits/bb0c02893475e03e46ef653f6886b90cc34cf7a9). So I am new to this syntax😂

From my understanding, it acts `Subalgebra.toSubring` on both parts of the equality `hcenter`. Thus `congr(Subalgebra.toSubring $hcenter)` yields `(Subalgebra.center L D).toSubring = ⊥.toSubring`, kinda like `apply_fun` in equality.

So here we can also close the target via
```
  have ntrivial : center D ≠ ⊤ := by
    apply_fun Subalgebra.toSubring at hcenter
    exact hcenter.trans_ne (Subalgebra.toSubring_injective.ne hneq)
```
or
```
  have ntrivial : center D ≠ ⊤ := by
    apply congrArg Subalgebra.toSubring at hcenter
    exact hcenter.trans_ne (Subalgebra.toSubring_injective.ne hneq)
```
Though I don't quite understand the 
>This is able to make use of the expected type, for example `(congr(_ + $h) : 1 + _ = _)` with `h : x = y` gives `1 + x = 1 + y`.


in its documentation.